tensorflow/python/eager/forwardprop.py

```diff
@@ -140,6 +140,48 @@ def _jvp_helper(op_name, attr_tuple, inputs, outputs, tangents):
     return output_tangents
 
 
+def _jvp_helper_wrapper(
+    op_name, attr_tuple, inputs, outputs, tangents, batch_size
+):
+    """Computes a batch of Jacobian-vector product for an op.
+
+  Args:
+    op_name: A string, the type of operation being executed.
+    attr_tuple: Attributes of the operation.
+    inputs: A flat list of input Tensors to the operation.
+    outputs: A flat list of output Tensors from the operation.
+    tangents: A flat list of Tensors, same shape as `[batch_size] + input_shape`.
+
+  Returns:
+    A flat list of tangents corresponding to `outputs`.
+  """
+  use_pfor = False
+  if batch_size is not None:
+    use_pfor = True
+    for primal, tangent in zip(inputs, tangents):
+      if tangent.rank == primal.rank + 1:
+        if tangent.shape != [batch_size] + primal.shape:
```

Comment:
Fine not to worry about it for now, but one thing to think about is dynamic shapes:

```python
import tensorflow as tf

@tf.function(input_signature=[tf.TensorSpec(shape=[], dtype=tf.float32),
                              tf.TensorSpec(shape=None, dtype=tf.float32)])
def f(batch_size, x):
  [batch_size] + x.shape  # error; tf.concat([[batch_size], tf.shape(x)], axis=0) works

f(tf.constant(16), tf.constant(1.))
```

I assume we'll want to allow dynamic batch sizes and undefined tangent shapes at some point.

Reply:
Made the change.